pkgs/development/compilers/llvm/common/default.nix

```diff
@@ -383,126 +393,103 @@ let
     }
     // lib.optionalAttrs (lib.versionAtLeast metadata.release_version "20") {
-      flang = callPackage ./flang {
-        mlir = tools.mlir;
-      };
-    }
-  );
+      flang = callPackage ./flang { };
 
-  libraries = lib.makeExtensible (
-    libraries:
-    let
-      callPackage = newScope (libraries // buildLlvmTools // args // metadata);
-    in
-    (
-      {
-        compiler-rt-libc = callPackage ./compiler-rt (
-          let
-            # temp rename to avoid infinite recursion
-            stdenv =
-              # Darwin needs to use a bootstrap stdenv to avoid an infinite recursion when cross-compiling.
-              if args.stdenv.hostPlatform.isDarwin then
-                overrideCC darwin.bootstrapStdenv buildLlvmTools.clangWithLibcAndBasicRtAndLibcxx
-              else if args.stdenv.hostPlatform.useLLVM or false then
-                overrideCC args.stdenv buildLlvmTools.clangWithLibcAndBasicRtAndLibcxx
-              else
-                args.stdenv;
-          in
-          {
-            inherit stdenv;
-          }
-        );
+      libc-overlay = callPackage ./libc {
+        isFullBuild = false;
+        # Use clang due to "gnu::naked" not working on aarch64.
+        # Issue: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=77882
+        stdenv = overrideCC stdenv buildLlvmPackages.clang;
+      };
 
-        compiler-rt-no-libc = callPackage ./compiler-rt {
-          doFakeLibgcc = stdenv.hostPlatform.useLLVM or false;
-          stdenv =
-            # Darwin needs to use a bootstrap stdenv to avoid an infinite recursion when cross-compiling.
-            if stdenv.hostPlatform.isDarwin then
-              overrideCC darwin.bootstrapStdenv buildLlvmTools.clangNoLibcNoRt
-            else
-              overrideCC stdenv buildLlvmTools.clangNoLibcNoRt;
-        };
-
-        compiler-rt =
-          if
-            stdenv.hostPlatform.libc == null
-            # Building the with-libc compiler-rt and WASM doesn't yet work,
-            # because wasilibc doesn't provide some expected things. See
-            # compiler-rt's file for further details.
-            || stdenv.hostPlatform.isWasm
-            # Failing `#include <term.h>` in
-            # `lib/sanitizer_common/sanitizer_platform_limits_freebsd.cpp`
-            # sanitizers, not sure where to get it.
-            || stdenv.hostPlatform.isFreeBSD
-          then
-            libraries.compiler-rt-no-libc
+      libc-full = callPackage ./libc {
+        isFullBuild = true;
+        # Use clang due to "gnu::naked" not working on aarch64.
+        # Issue: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=77882
+        stdenv = overrideCC stdenv buildLlvmPackages.clangNoLibcNoRt;
+        cmake =
+          if stdenv.targetPlatform.libc == "llvm" then buildPackages.cmakeMinimal else buildPackages.cmake;
+        python3 =
+          if stdenv.targetPlatform.libc == "llvm" then
+            buildPackages.python3Minimal
```

Comment:
(This explicit `buildPackages` use is a bit screwy – these are used in `nativeBuildInputs` anyway, so this just breaks splicing for seemingly no reason. However, this is a pre‐existing problem, of course; I’m just noting it here as I pore over the cross stuff here.)